### PR TITLE
[Feat][Pool] Implement MaxPool1d and MaxPool3d forward operators

### LINKED ISSUE #1726: [Feat][Ops] MaxPool1d/MaxPool3d: kernel/op implementation, tests, and benchmarks
## Parent Issue

Part of #1672

## Task Type

- [ ] **Manifest / Spec** (required for any public Op)
- [x] **Kernel / Op Implementation**
- [x] **Correctness Tests / Workloads**
- [x] **Benchmark / Performance**

## Description

Implement the four remaining spec-only MaxPool operators declared in `tileops/manifest/pool.yaml`, following the merged MaxPool2d implementation (#1722) as the structural template:

- `MaxPool1dFwdOp` / `MaxPool1dIndicesFwdOp` — NCL layout `(N, C, L_in)`, reference `torch.nn.functional.max_pool1d`
- `MaxPool3dFwdOp` / `MaxPool3dIndicesFwdOp` — NCDHW layout `(N, C, D_in, H_in, W_in)`, reference `torch.nn.functional.max_pool3d`

Scope:

- Kernels: new `tileops/kernels/pool/max_pool1d.py` and `tileops/kernels/pool/max_pool3d.py`, each with a values kernel and a with-indices kernel (int64 per-(N,C)-plane flat indices, NaN semantics aligned with MaxPool2d), wired through `kernel_map` dispatch and exported from the kernels `__init__` files.
- Ops: add `MaxPool1dFwdOp` / `MaxPool1dIndicesFwdOp` / `MaxPool3dFwdOp` / `MaxPool3dIndicesFwdOp` to `tileops/ops/pool.py` with input-inferred shape metadata (per #1648), `_infer_output_shapes`, `eval_roofline`, and dtype validation; export from `tileops/ops/__init__.py`.
- Manifest: promote the four entries in `tileops/manifest/pool.yaml` from `spec-only` to `implemented` and fill in `source` (kernel, kernel_map, op, test, bench, `bench_manifest_driven`).
- Tests: extend `tests/ops/test_pool.py` with fixture-based correctness tests covering all dtypes (fp16/bf16/fp32) in smoke, representative stride/padding/dilation/ceil_mode/stride=None cases in full, plus NaN / all-`-inf` window edge cases for values and indices semantics.
- Benchmarks: extend `benchmarks/ops/bench_pool.py` with manifest-workload-driven benchmarks recording `tileops` vs `torch-ref` baselines for both values and indices variants.

## Checklist

- [ ] Public Op has a `tileops/manifest/` entry (or updates an existing entry).
- [ ] Manifest `signature` declares inputs, outputs, params, shape rules, and dtype coverage.
- [ ] Manifest `workloads` declare benchmark shapes/dtypes; unit-test edge cases are not generated from manifest workloads.
- [ ] Manifest `roofline` is present and consumable by `op.eval_roofline()`.
- [ ] Manifest `source` declares kernel, op, test, bench, and `source.kernel_map` when dispatching kernels.
- [ ] Op constructor, `forward()`, and `default_kernel_map` match the manifest entry.
- [ ] Tests use an independent reference implementation and cover relevant FP16/BF16 and edge cases.
- [ ] Benchmarks consume manifest workloads and record at least one non-`tileops` baseline unless explicitly justified.
- [ ] PR title and commits follow the current `[Feat][Scope]` / `[Perf][Scope]` / `[Fix][Scope]` convention.
- [ ] Implementation follows **Google Python Style** for code and docstrings.

### PR DESCRIPTION
**Pull request by RMLYC**
Closes #1726

## Summary

- implement MaxPool1d and MaxPool3d values/with-indices TileLang kernels for NCL and NCDHW inputs
- add input-inferred Ops, kernel dispatch/cache, roofline evaluation, exports, and manifest source metadata
- add PyTorch-referenced correctness coverage for fp16/bf16/fp32, stride/padding/dilation/ceil mode, non-contiguous inputs, NaN, all-`-inf`, tied maxima, and flattened int64 indices
- add manifest-driven benchmarks for values and indices variants

## Test plan

- [x] pre-commit hooks passed
- [x] `python scripts/validate_manifest.py --check-op MaxPool1dFwdOp`
- [x] `python scripts/validate_manifest.py --check-op MaxPool1dIndicesFwdOp`
- [x] `python scripts/validate_manifest.py --check-op MaxPool3dFwdOp`
- [x] `python scripts/validate_manifest.py --check-op MaxPool3dIndicesFwdOp`
- [x] MaxPool1d smoke/full: 42 passed
- [x] MaxPool3d smoke/full: 42 passed
- [x] pool smoke regression: 83 passed
- [x] MaxPool1d/3d benchmarks: 12 passed

## Benchmark

Measured on NVIDIA H200 (SM90), PyTorch 2.10.0+cu129. The acceptance threshold is TileOPs performance at least 90% of PyTorch, equivalent to `torch-ref latency / tileops latency >= 0.9`.

| Variant / workload | TileOPs (ms) | PyTorch (ms) | PyTorch / TileOPs |
|---|---:|---:|---:|
| MaxPool1d / sincnet-speaker-local fp16 | 0.0117 | 0.0513 | 4.38x |
| MaxPool1d / textcnn-global fp16 | 0.0141 | 0.0271 | 1.92x |
| MaxPool1d / ecg-cnn-dilated bf16 | 0.0115 | 0.0362 | 3.15x |
| MaxPool1d indices / sincnet-speaker-local fp16 | 0.0226 | 0.0513 | 2.27x |
| MaxPool1d indices / textcnn-global fp16 | 0.0197 | 0.0271 | 1.38x |
| MaxPool1d indices / ecg-cnn-dilated bf16 | 0.0214 | 0.0362 | 1.69x |
| MaxPool3d / c3d-pool1 fp16 | 0.0753 | 0.4932 | 6.55x |
| MaxPool3d / c3d-pool2 fp16 | 0.0244 | 0.0868 | 3.56x |
| MaxPool3d / medicalnet-stem bf16 | 0.1314 | 0.3340 | 2.54x |
| MaxPool3d indices / c3d-pool1 fp16 | 0.3050 | 0.4935 | 1.62x |
| MaxPool3d indices / c3d-pool2 fp16 | 0.0593 | 0.0865 | 1.46x |
| MaxPool3d indices / medicalnet-stem bf16 | 0.3157 | 0.3333 | 1.06x |

All 12 workloads pass the performance requirement. Values kernels are 1.92x-6.55x faster than PyTorch. Indices variants narrow the gap to 1.06x-2.27x because they additionally track selection state and write one int64 index per output element; the largest relative cost appears on the bandwidth-heavy MedicalNet 3D workload, which still remains 1.06x faster than PyTorch.

### Why these performance results are reasonable

The two largest gaps were independently reproduced with Nsight Systems using the autotuned fixed configs and exact-output validation (`atol=0, rtol=0`):

| Representative case | TileOPs nsys median | PyTorch nsys median | Ratio | PyTorch / TileOPs executed instructions |
|---|---:|---:|---:|---:|
| MaxPool1d sincnet-speaker-local `[32,80,4096]`, k3s3 fp16 | 11.68 us | 51.23 us | 4.39x | 10.7x |
| MaxPool3d c3d-pool1 `[8,64,16,112,112]`, k(1,2,2) fp16 | 75.07 us | 493.09 us | 6.57x | 8.8x |

For the 1d values case, PyTorch dispatches the generic `max_pool_forward_nchw<Half, int>(..., Half*, long*)` kernel. Nsight Compute reports 31.67M executed instructions versus 2.95M for the dimension-specialized TileOPs kernel. The `long*` output also shows that the PyTorch values path retains indices-tracking work even though the public call uses `return_indices=False`.

For the 3d values case, PyTorch dispatches `max_pool3d_with_indices_single_out_frame<Half>(..., Half*, long*, ...)`. It executes 317.46M instructions versus 35.93M for TileOPs and writes about 240.48 MB to DRAM versus 46.17 MB. The hidden int64 indices tensor is 205.52 MB for this shape. TileOPs avoids that hidden output and reaches 74.05% DRAM throughput (3.64 TB/s), while the generic PyTorch kernel reaches 18.96%.

The benchmark comparison is API-level and symmetric: both implementations receive the same shape, dtype and pooling parameters and are checked against the same `return_indices=False` result. `BenchmarkBase.profile` measures GPU kernel time only, after 10 warmups, with 50 repeats across 3 trials, L2 flushes and rotating input clones; JIT, autotune and Python overhead are outside the timed region. Nsight Systems independently reproduces the reported ratios with low variance. These ratios characterize the PyTorch 2.10 H200 CUDA backend and should not be generalized to every GPU, PyTorch version, or a hypothetical values-only specialized PyTorch kernel.

## Additional context

- follows the merged MaxPool2d implementation in #1722 as the structural template
- preserves PyTorch semantics: ties select the first maximum, while windows containing NaNs return NaN and record the last visited NaN index

### CODE REVIEW

File: tileops/kernels/pool/max_pool3d.py [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+                                    ):
+                                        val = T.cast(x[batch, c_idx, id_, ih, iw], accum_dtype)
+                                        has_nan = has_nan | T.isnan(val)
+                                        max_val = T.max(max_val, val)
+
+                        result = T.if_then_else(
+                            has_nan,
+                            T.cast(float("nan"), accum_dtype),
+                            max_val,
+                        )
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tileops/kernels/pool/max_pool3d.py:90]: Addressed in commit 94d2e9c8: replaced the boolean bitwise-OR accumulation with an explicit conditional (if T.isnan(val): has_nan = True) in the MaxPool3d values kernel. Validation: 42 MaxPool3d smoke/full tests passed, all 12 MaxPool1d/3d benchmarks passed, and c3d-pool1 remained at 0.0763 ms (within about 1.3% of the previous run).

File: tileops/kernels/pool/max_pool1d.py [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+
+                        max_val = T.alloc_var(T.float32)
+                        has_nan = T.alloc_var(T.bool)
+                        max_val = T.cast(float("-inf"), accum_dtype)
+                        has_nan = False
+                        for kw in T.serial(kernel_w):
+                            iw = ow * stride_w - pad_w + kw * dilation_w
+                            if iw >= 0 and iw < l_in:
+                                val = T.cast(x[batch, c_idx, iw], accum_dtype)
+                                has_nan = has_nan | T.isnan(val)
```
[the remark this answers is not in the document]
Comment (RMLYC) [reply tileops/kernels/pool/max_pool1d.py]: Addressed in commit 94d2e9c8: replaced the boolean bitwise-OR accumulation with an explicit conditional (if T.isnan(val): has_nan = True) in the MaxPool1d values kernel. Validation: 42 MaxPool1d smoke/full tests passed, all 12 MaxPool1d/3d benchmarks passed, and sincnet-speaker-local remained at 0.0117 ms.

**Review by superAngGao — APPROVED** [review]
LGTM. I checked the MaxPool1d/3d op/kernel paths against the existing MaxPool2d structure, including forward-time shape/dtype binding, kernel cache keys, PyTorch-compatible NaN/tie/indices behavior, manifest source metadata, exports, and benchmark wiring. The earlier boolean-or concern has been addressed by the follow-up commit, and the current CI checks are green.

**Review by zhen8838 — APPROVED** [review]
Reviewed current head 94d2e9c8599d24025be581cb5ca49afbe52fea0c.

MaxPool1d/3d values and indices paths are consistent across parameter validation, ceil/dilation/padding semantics, NaN and tie handling, flattened indices, custom-op fake registrations, exports, and manifest source metadata. No blocking correctness or contract issue found.

Relevant CI is green, including gpu-smoke, pre-commit, ci-gate, validate-pr-title, and manifest stats checks.
